toolchain/semantics/semantics_context.cpp

```diff
@@ -448,6 +448,44 @@ auto SemanticsContext::ParamOrArgSave(bool for_args) -> void {
   params_or_args.push_back(entry_node_id);
 }
 
+template <typename ProfileType, typename MakeNode>
+auto SemanticsContext::CanonicalizeTypeImpl(SemanticsNodeKind kind,
+                                            ProfileType profile_type,
+                                            MakeNode make_node)
+    -> SemanticsTypeId {
+  llvm::FoldingSetNodeID canonical_id;
+  kind.Profile(canonical_id);
+  profile_type(canonical_id);
+
+  void* insert_pos;
+  auto* node =
+      canonical_type_nodes_.FindNodeOrInsertPos(canonical_id, insert_pos);
+  if (node != nullptr) {
+    return node->type_id();
+  }
+
+  auto node_id = make_node();
+  auto type_id = semantics_ir_->AddType(node_id);
+  CARBON_CHECK(canonical_types_.insert({node_id, type_id}).second);
+  type_node_storage_.push_back(
+      std::make_unique<TypeNode>(canonical_id, type_id));
+
+#ifndef NDEBUG
+  // In a debug build, check that our insertion position is still valid. It
+  // could have been invalidated by a misbehaving `make_node`.
+  void* check_insert_pos;
+  auto* check_node =
+      canonical_type_nodes_.FindNodeOrInsertPos(canonical_id, check_insert_pos);
+  CARBON_CHECK(!check_node)
+      << "Type was created recursively during canonicalization";
+  CARBON_CHECK(insert_pos == check_insert_pos)
+      << "Insertion position changed during canonicalization";
```

Comment:
Per discussion -- you might consider introducing DCHECK, particularly since it'd allow dropping NDEBUG (personally, I'd write that Find in the DCHECK simply because I don't like #ifdef inside code)

Reply:
Done.